### PR TITLE
ATM: Test for endpoints scored at inference time

### PR DESCRIPTION
**Pull request by tiferet**
Adds a test to detect changes in the endpoints that get scored at inference time.

Note that the queries' `.ql` files (e.g. `src/NosqlInjectionATM.ql`) can't be called directly here, because they use the model and compute a score.

Closes https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2135

### CODE REVIEW

**Review by jhelie — APPROVED** [review]
Thanks @tiferet, LGTM. I've added `XssThroughDom` but this didn't require updating the `.expected` files : I guess our some small examples do not contain instances of source for `Xss` and not for `XssThroughDom` (or vice versa).

**Comment by jhelie** [comment]
ah I see we were both working on the PR: I can push my change or you can update the `ExtractEndpointDataInference.ql` file yourself

**Comment by tiferet** [comment]
> Thanks @tiferet, LGTM. I've added `XssThroughDom` but this didn't require updating the `.expected` files : I guess our some small examples do not contain instances of source for `Xss` and not for `XssThroughDom` (or vice versa).

I don't see any commits from you, but don't worry about it -- I'll make the needed 1-line change and push it 😄

File: javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ExtractEndpointDataInference.ql
Comment (kaeluka) [inline javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ExtractEndpointDataInference.ql:24]:
```suggestion
  AtmConfig::AtmConfig queryConfig, JS::DataFlow::Node sink
```

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll
Comment (kaeluka) [inline javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:162]:
```suggestion
  predicate isSinkCandidateWithFlow(JS::DataFlow::Node sink) {
```

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll
Comment (kaeluka) [inline javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:164]:
```suggestion
      this.hasFlow(source, sink) and
```

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll
Comment (kaeluka) [inline javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:165]:
```suggestion
      not AtmResultsInfo::isFlowLikelyInBaseQuery(source, sink)
```

File: javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ExtractEndpointDataInference.ql
Comment (kaeluka) [inline javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ExtractEndpointDataInference.ql:26]:
```suggestion
  queryConfig.isSinkCandidateWithFlow(sink)
```

**Review by kaeluka — COMMENTED** [review]
LGTM, but I think we don't need to rely on the heavier `PathNode` class here. `Node` suffices.

Also suggested a name change to explicitly mention that the predicate only returns endpoints WITH FLOW.

**Comment by kaeluka** [comment]
> Does it make sense that there are no XssThroughDom sink candidates with flow from a source in endpoint_large_scale?

No, IMO

> Was that set created specifically for our existing four queries?

No, IMO (but I wasn't there)

> If so, do we need to add to it each time we boost a new query?

I think so, yes.

**Comment by jhelie** [comment]
Please update the [issue template](https://github.com/github/ml-ql-adaptive-threat-modeling/blob/main/.github/ISSUE_TEMPLATE/atm-add-boosted-query-using-existing-sink.md) if we need to consider updating `endpoint_large_scale` when adding a new query.

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:164]: I think this change is incorrect. We need to use `hasFlowPath` here, just like we use it in `hasBoostedFlowPath` above, so that the test is as close as possible to the actual extraction queries.

File: javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ExtractEndpointDataInference.ql
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ExtractEndpointDataInference.ql:24]: I want to keep the test as similar as possible to the actual extraction queries. The extraction queries use `DataFlow::PathNode` (e.g. `javascript/ql/experimental/adaptivethreatmodeling/src/SqlInjectionATM.ql`). I don't know if there's a reason they do this or not, but if we want to change those to `DataFlow::Node` (in which case we can change these as well), we should do so in a separate PR.

**Comment by tiferet** [comment]
> Please update the [issue template](https://github.com/github/ml-ql-adaptive-threat-modeling/blob/main/.github/ISSUE_TEMPLATE/atm-add-boosted-query-using-existing-sink.md) if we need to consider updating `endpoint_large_scale` when adding a new query.

@jhelie I added a line about it [here](https://github.com/github/ml-ql-adaptive-threat-modeling/blob/main/.github/ISSUE_TEMPLATE/atm-add-boosted-query-using-existing-sink.md#step-1-create-new-codeql-files-in-githubcodeql).

**Comment by henrymercer** [comment]
> @kaeluka / @henrymercer Does it make sense that there are no XssThroughDom sink candidates with flow from a source in `endpoint_large_scale`? Was that set created specifically for our existing four queries? If so, do we need to add to it each time we boost a new query?

Stephan is correct, and I'll add some more context.  See https://github.com/github/codeql/tree/main/javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/autogenerated for a description of how `endpoint_large_scale` is generated.  Ideally, the files copied from `javascript/ql/test/query-tests/Security/CWE-079` would contain some sink candidates with flow, but it looks like the model isn't finding anything new for XSS through DOM.

The test set was not created specifically for the existing four queries, but in general we will need to check it whenever we boost a new query to ensure it covers the new query.

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll
Comment (kaeluka) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:164]: I'll have to look into what the difference is on Monday morning. I think it's correct, but that's not good enough if you disagree.

File: javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ExtractEndpointDataInference.ql
Comment (kaeluka) [reply javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ExtractEndpointDataInference.ql:24]: The extraction queries use the PathNodes for a specific reason - namely, that the UI should be able to list a specific path from source to sink. This is not needed in this test.

**Comment by tiferet** [comment]
> > @kaeluka / @henrymercer Does it make sense that there are no XssThroughDom sink candidates with flow from a source in `endpoint_large_scale`? Was that set created specifically for our existing four queries? If so, do we need to add to it each time we boost a new query?
> 
> Stephan is correct, and I'll add some more context. See https://github.com/github/codeql/tree/main/javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/autogenerated for a description of how `endpoint_large_scale` is generated. Ideally, the files copied from `javascript/ql/test/query-tests/Security/CWE-079` would contain some sink candidates with flow for XSS through DOM, but it looks like the model isn't finding anything new there.
> 
> The test set was not created specifically for the existing four queries, but in general we will need to check it whenever we boost a new query to ensure it covers the new query.

Thanks! @jhelie I linked this answer in the issue template as well, for when you get back to the XssThroughDom work.

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:164]: The question is what's the harm in leaving both this and `PathNode` as they are, so that they match the extraction queries as closely as possible. If there's no major downside, I think that's the right call. 

Ideally the tests would call the extraction queries directly, to guarantee we're testing exactly what we intend to test, but in this case we can't do that because the extraction queries use the ML model. So the idea was to do everything exactly like the extraction queries other than producing a score. I understand we may be able to use `Node` rather than `PathNode` and `hasFlow` rather than `hasFlowPath`, but I don't understand what benefit we'd get from that that's more important than ensuring the test matches the extraction queries as closely as possible.

File: javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ExtractEndpointDataInference.ql
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ExtractEndpointDataInference.ql:24]: See my comment [here](https://github.com/github/codeql/pull/11532#discussion_r1038366721)

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll
Comment (kaeluka) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:164]: I thought about this some more over dinner and came to the conclusion that it's probably not worth holding this up over something that is very likely only a tiny performance issue. 👍

Comment (kaeluka) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:164]: ...so I walked upstairs to tell you that. I'll now look at the discussions you and Jean had in this PR and if there's nothing left open from him, will approve.

**Review by kaeluka — APPROVED** [review]
The PR LGTM, although I'm a bit unclear whether the discussion between @tiferet and Jean has been resolved already.
 - I'm approving this, assuming that all things related to the conversation will be resolved in a different PR. This is how I understood the conversation.
 - The discussion about PathNodes is not worth losing time over (but it somewhat has bumped the urgency with which I want to look into that class's implementation ^^)

Also: thanks, @henrymercer for weighing in! I never had to look "inside" those tests before, so I appreciate the background info. I actually thought they were hand-crafted for the individual queries.

**Comment by tiferet** [comment]
> The PR LGTM, although I'm a bit unclear whether the discussion between @tiferet and Jean has been resolved already.

That was actually a conversation about the addition of XssThroughDom, that ended up here just because this PR revealed that our test set lacks XssThroughDom examples 😄. That's part of the XssThroughDom work, though, unrelated to this PR.

**Pull request merged**
